ruby/ql/lib/codeql/ruby/ast/internal/Constant.qll

```diff
@@ -509,3 +509,23 @@ private module Cached {
 }
 
 import Cached
+
+/**
+ * Holds if `e` is an array constructed from an array literal.
+ * Example:
+ * ```rb
+ * [1, 2, 3]
+ * C = [1, 2, 3]; C
+ * x = [1, 2, 3]; x
+ * ```
+ */
+predicate isArrayConstant(ExprCfgNode e, ArrayLiteralCfgNode arr) {
```

Comment:
This is a bit less general than my proposal, which would handle (probably not likely in practice) cases like
```
x = [1, 2, 3]; C = x; y = C; y
```

Reply:
Tom's example doesn't make sense to me either. I *think* he meant to write this:

```codeql
predicate isArray(ExprCfgNode e, ArrayLiteralCfgNode a) {
  e = a
  or
  isArrayExpr(e.getExpr(), a)
  or
  isArray(getSource(e), a)
}

predicate isArrayExpr(Expr e, ArrayLiteralCfgNode a) {
  isArrayExpr(e.(ConstantReadAccess).getValue(), a)
  or
  forex(ExprCfgNode n | n = e.getAControlFlowNode() | isArray(n, a))
}
```

If you're not feeling confident with that change, I still think it's worth merging your version of `isArrayConstant`, but perhaps with a TODO comment including the example Tom gave with the series of assignments your version won't handle.